src/main/java/org/opensearch/security/http/OnBehalfOfAuthenticator.java

```diff
@@ -0,0 +1,254 @@
+        });
+
+        return creds;
+    }
+
+    private AuthCredentials extractCredentials0(final RestRequest request) {
+        if (!oboEnabled) {
+            log.error("On-behalf-of authentication has been disabled");
+            return null;
+        }
+
+        if (jwtParser == null) {
+            log.error("Missing Signing Key. JWT authentication will not work");
+            return null;
+        }
+
+        String jwtToken = request.header(HttpHeaders.AUTHORIZATION);
+
+        if (jwtToken == null || jwtToken.length() == 0) {
+            if (log.isDebugEnabled()) {
+                log.debug("No JWT token found in '{}' header", HttpHeaders.AUTHORIZATION);
+            }
+            return null;
+        }
+
+        if (!BEARER.matcher(jwtToken).matches()) {
+            jwtToken = null;
+        }
+
+        if (jwtToken != null && Pattern.compile(BEARER_PREFIX).matcher(jwtToken.toLowerCase()).find()) {
+            jwtToken = jwtToken.substring(jwtToken.toLowerCase().indexOf(BEARER_PREFIX) + BEARER_PREFIX.length());
+        } else {
+            if (log.isDebugEnabled()) {
+                log.debug("No Bearer scheme found in header");
+            }
+        }
+
+        if (jwtToken == null) {
+            return null;
+        }
+
+        try {
+            if (!isAllowedRequest(request)) {
+                return null;
+            }
+
+            final Claims claims = jwtParser.parseClaimsJws(jwtToken).getBody();
+
+            final String subject = claims.getSubject();
+            if (Objects.isNull(subject)) {
+                log.error("Valid jwt on behalf of token with no subject");
+                return null;
+            }
+
+            final String audience = claims.getAudience();
+            if (Objects.isNull(audience)) {
+                log.error("Valid jwt on behalf of token with no audience");
+                return null;
+            }
+
+            final String issuer = claims.getIssuer();
+            if (!issuer.equals(clusterName)) {
+                log.error("This issuer of this OBO does not match the current cluster identifier");
+                return null;
+            }
+
+            List<String> roles = extractSecurityRolesFromClaims(claims);
+            String[] backendRoles = extractBackendRolesFromClaims(claims);
+
+            final AuthCredentials ac = new AuthCredentials(subject, roles, backendRoles).markComplete();
+
+            for (Entry<String, Object> claim : claims.entrySet()) {
+                ac.addAttribute("attr.jwt." + claim.getKey(), String.valueOf(claim.getValue()));
+            }
+
+            return ac;
+
+        } catch (WeakKeyException e) {
+            log.error("Cannot authenticate user with JWT because of ", e);
+            return null;
+        } catch (Exception e) {
+            e.printStackTrace();
+            if (log.isDebugEnabled()) {
+                log.debug("Invalid or expired JWT token.", e);
+            }
+            return null;
```

Comment:
can extract this and the one above to a finally block:
```
...
} finally {
   return null;
}
```

Reply:
If we do the extraction like the following:
```
try {
    // ... (some code)
    return ac;
} catch (WeakKeyException e) {
    log.error("Cannot authenticate user with JWT because of ", e);
} catch (Exception e) {
    if (log.isDebugEnabled()) {
        log.debug("Invalid or expired JWT token.", e);
    }
} finally {
   return null;
}
```
I think if we use this structure for the to `catch`es, even when the code in the `try` block completes successfully and prepares to return `ac`, the `finally` block will still execute and will return `null`, overriding the previously intended return value.